### DISCUSSION TITLE
Proposal to Integrate a New Cryptographic Algorithm into OpenSSL

### DISCUSSION THREAD

**Discussion by yaz2109** [openssl/openssl#25062] [Q&A] [answered]
Hello OpenSSL team,

My name is Yaswanth a B. Tech. graduand from India, and I have developed a new cryptographic algorithm in Python that I believe could be a valuable addition to the OpenSSL project. I am writing to discuss the possibility of integrating this algorithm into OpenSSL for which I seek guidance over the procedure required to achieve this. I am in the process of translating this into C/CPP for compatibility with OpenSSL.

I would like to request guidance on the following:

1. **Review Process:** What is the process for submitting a new cryptographic algorithm for review by the OpenSSL team? And will there be any possibility where I can fix a key-size for a specific device category like wearables?
2. **Implementation Guidelines:** Are there specific guidelines or best practices for implementing cryptographic algorithms in OpenSSL?
3. **Testing and Validation:** What are the requirements for testing and validating the algorithm to ensure it meets OpenSSL's standards?
4. **Documentation:** What level of documentation is required to accompany the implementation?

I am excited about the possibility of contributing to OpenSSL and look forward to your feedback and guidance. I am ready to work closely with the OpenSSL team to ensure that the proposed algorithm is implemented as an open-source deliverable.
Thank you for your time and consideration.

Best regards,
Yaswanth S.

**Comment by mattcaswell** [comment, marked as answer]
You don't tell us anything about what type of cryptographic algorithm you are working with (encryption, signature, mac, kdf, etc). However in all cases we will only integrate a new cryptographic algorithm into OpenSSL if it has been standardised by some recognised standards body (e.g. IETF, NIST, ISO etc). We will also probably want to see some evidence of real world use likely to add benefit to our users.

So - the first question is - has the algorithm been standardised? If not then your best option is to not integrate it into OpenSSL itself at all. Instead you can write a third party provider as a plug-in to OpenSSL. The OpenSSL provider architecture was designed to enable third parties to add whatever algorithms they like into OpenSSL. You can do that entirely by yourself in your own repository without having to adhere to OpenSSL review processes.

See the following links for some resources on writing a provider:

https://docs.openssl.org/master/man7/provider/
https://www.youtube.com/watch?v=mHrB1RKlkYk
https://www.youtube.com/watch?v=ztfbzsnHqKU

**Comment by mattcaswell** [comment]
Of course I should add the warning about rolling your own crypto...

https://security.stackexchange.com/questions/18197/why-shouldnt-we-roll-our-own
